xilem/src/style.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2025 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Traits used to set custom styles on elements.
+
+use masonry_winit::core::Property;
+use masonry_winit::properties::types::Gradient;
+use masonry_winit::properties::{
+    Background, BorderColor, BorderWidth, BoxShadow, CornerRadius, Padding,
+};
+use vello::peniko::Color;
+
+use crate::property_tuple::PropertyTuple;
+
+/// Marker trait implement by elements to signal that a given property can be set on them.
+pub trait HasProperty<P: Property>: Style {
+    /// Return a mutable reference to the specific property.
+    fn property(&mut self) -> &mut Option<P>;
+}
+
+/// Trait implemented by most elements that lets you set some styling properties.
+///
+/// Which methods you can use will depend on which marker traits the element implements,
+/// which matches which [`Properties`](masonry_winit::core::Properties) the underlying widget handles.
+pub trait Style: Sized {
+    /// The tuple type used by the element to store properties.
+    type Props: PropertyTuple;
+
+    /// Return a mutable reference to the element's property storage.
+    fn properties(&mut self) -> &mut Self::Props;
```

Comment:
If I were designing this, I probably wouldn't have had this, and instead provided the tuple field to the macro. But I won't block on that; there are tradeoffs either way.

Reply:
Overall I want to do as little as I can inside the macro. Macros are bad DX in general. Compare:

```rust
crate::declare_property_tuple!(
    ButtonProps;
    Button<F>;
    self.properties;

    Background, 0;
    BorderColor, 1;
    BorderWidth, 2;
    BoxShadow, 3;
    CornerRadius, 4;
    Padding, 5;
);
```

with

```rust
impl<F> Style for Button<F> {
    type Props = ButtonProps;

    fn properties(&mut self) -> &mut Self::Props {
        &mut self.properties
    }
}
```

Someone reading the code will have an easier time understanding what `self.properties` means and why it's there in the trait version. Error messages will be clearer, etc.